### PR TITLE
[Plugins] - MON-16256 Retrieve alerts from kadiska

### PR DESCRIPTION
**Pull request by lucie-dubrunfaut**
## Description

Retrieve rules alerts from Kadiska with new mode alerts.
New discovery service for rules alerts.

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

<h2> How this pull request can be tested ? </h2>

To test this PR you need access to Kadiska credentials (client-id and client-secret)
Also you need to set up alert rules 

Example how to test alerts.pm : (period=120 minutes => 2 hours)
```
perl centreon_plugins.pl     --plugin=apps::monitoring::kadiska::plugin --mode=alerts --client-id=CLIENT_ID --client-secret=CLIENT_SECRET --period=120 --verbose
```

Example how to test listrules.pm : (period=120 minutes => 2 hours)
```
perl centreon_plugins.pl     --plugin=apps::monitoring::kadiska::plugin --mode=list-rules --client-id=CLIENT_ID --client-secret=CLIENT_SECRET --period=120 --verbose
```

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

**Review by sdepassio — APPROVED** [review]

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+            }
+        },
+        { label => 'rules-warnings-total-count', nlabel => 'warnings.total.count',
+            set => {
+                key_values => [ { name => 'total_warning' }  ],
+                output_template => 'total warnings: %s',
+                perfdatas => [ { template => '%d', min => 0 } ]
+            }
+        },
+        { label => 'rules-nodata-total-count', nlabel => 'no.data.total.count',
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:71]: the nlabel: ```rules.nodata.count```

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    ];
+
+    $self->{maps_counters}->{global} = [
+        { label => 'rules-total-count', nlabel => 'rules.total.count', set => {
+                key_values => [ { name => 'total' }  ],
+                output_template => 'total rules: %s',
+                perfdatas => [ { template => '%d', min => 0 } ]
+            }
+        },
+        { label => 'rules-criticals-total-count', nlabel => 'criticals.total.count',
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:56]: the nlabel: ```rules.criticals.count```

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+            set
+                => {
+                key_values => [ { name => 'total_critical' }  ],
+                output_template => 'total criticals: %s',
+                perfdatas => [ { template => '%d', min => 0 } ]
+            }
+        },
+        { label => 'rules-warnings-total-count', nlabel => 'warnings.total.count',
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:64]: the nlabel: ```rules.warnings.count```

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+                key_values => [ { name => 'total_nodata' }  ],
+                output_template => 'total no data: %s',
+                perfdatas => [ { template => '%d', min => 0 } ]
+            }
+        }
+    ];
+
+    $self->{maps_counters}->{rules} = [
+        { label => 'rule-ok-count',
+            type => 1,
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:83]: No need to set ```type => 1```. Why we don't have a perfdata for that counter ?

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+            set => {
+                key_values => [ { name => 'ok_count' } ],
+                output_template => 'ok count: %s',
+                display_ok => 0,
+                closure_custom_perfdata => sub { return 0; }
+            }
+        },
+        { label => 'rule-warning-count',
+            type => 1,
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:92]: No need to set ```type => 1```. Why we don't have a perfdata for that counter ?

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+            set => {
+                key_values => [ { name => 'critical_count' } ],
+                output_template => 'critical count: %s',
+                display_ok => 0,
+                closure_custom_perfdata => sub { return 0; }
+            }
+        },
+        { label => 'rule-nodata-count',
+            type => 1,
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:110]: No need to set ```type => 1```. Why we don't have a perfdata for that counter ?

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            set => {
+                key_values => [ { name => 'warning_count' } ],
+                output_template => 'warning count: %s',
+                display_ok => 0,
+                closure_custom_perfdata => sub { return 0; }
+            }
+        },
+        { label => 'rule-critical-count',
+            type => 1,
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:101]: No need to set ```type => 1```. Why we don't have a perfdata for that counter ?

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+    #Check if bad API request is submit
+    if (!exists $results->{data}) {
+        $self->{output}->add_option_msg(short_msg => 'No data result in API request.');
+        $self->{output}->option_exit();
+    }
+
+    $self->{total_critical}->{total_critical}=0;
+    $self->{total_warning}->{total_warning}=0;
+    $self->{total_nodata}->{total_nodata}=0;
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:185]: if think it's wrong. it should be:
```
$self->{global} = { total_critical => 0, total_warning => 0, total_nodata => 0 };
```

File: src/apps/monitoring/kadiska/plugin.pm [hunk 9]
```diff
@@ -30,11 +30,13 @@ sub new {
     bless $self, $class;
 
     $self->{modes} = {
-        'list-watchers'         => 'apps::monitoring::kadiska::mode::listwatchers',
+        'alerts'                => 'apps::monitoring::kadiska::mode::alerts',
+        'list-alertrules'       => 'apps::monitoring::kadiska::mode::listalertrules',
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/plugin.pm:34]: it should be better to use the name ```list-alerts```

**Review by garnier-quentin — COMMENTED** [review]
there are some indentation small problem also.

File: src/apps/monitoring/kadiska/plugin.pm (hunk 9 above)
Comment (omercier) [reply src/apps/monitoring/kadiska/plugin.pm:34]: I don't see it that way. This mode is about listing alert rules in order to create services that will count the number of alerts by alert rule.

Comment (garnier-quentin) [reply src/apps/monitoring/kadiska/plugin.pm:34]: so call it list-alert-rules

File: src/apps/monitoring/kadiska/mode/alerts.pm (hunk 5 above)
Comment (lucie-dubrunfaut) [reply src/apps/monitoring/kadiska/mode/alerts.pm:92]: This section was developped with this plugin as pattern :
https://github.com/centreon/centreon-plugins/blob/develop/src/hardware/devices/hms/netbiter/argos/restapi/mode/alarms.pm#L90

Does this use of perfdata is correct ?
```
perfdatas => [
        { template => '%s', min => 0, label_extra_instance => 1, instance_use => 'name' }
]
```

File: src/apps/monitoring/kadiska/mode/alerts.pm (hunk 7 above)
Comment (lucie-dubrunfaut) [reply src/apps/monitoring/kadiska/mode/alerts.pm:101]: This section was developped with this plugin as pattern :
https://github.com/centreon/centreon-plugins/blob/develop/src/hardware/devices/hms/netbiter/argos/restapi/mode/alarms.pm#L90

Does this use of perfdata is correct ?
```
perfdatas => [
        { template => '%s', min => 0, label_extra_instance => 1, instance_use => 'name' }
]
```

File: src/apps/monitoring/kadiska/mode/alerts.pm (hunk 6 above)
Comment (lucie-dubrunfaut) [reply src/apps/monitoring/kadiska/mode/alerts.pm:110]: This section was developped with this plugin as pattern :
https://github.com/centreon/centreon-plugins/blob/develop/src/hardware/devices/hms/netbiter/argos/restapi/mode/alarms.pm#L90

Does this use of perfdata is correct ?
```
perfdatas => [
        { template => '%s', min => 0, label_extra_instance => 1, instance_use => 'name' }
]
```

File: src/apps/monitoring/kadiska/mode/alerts.pm (hunk 7 above)
Comment (garnier-quentin) [reply src/apps/monitoring/kadiska/mode/alerts.pm:101]: Yes but you need to add the name in ```key_values```

File: src/apps/monitoring/kadiska/mode/alerts.pm (hunk 8 above)
Comment (lucie-dubrunfaut) [reply src/apps/monitoring/kadiska/mode/alerts.pm:185]: Both seems work but I add your suggestion.

File: src/apps/monitoring/kadiska/mode/alerts.pm (hunk 7 above)
Comment (lucie-dubrunfaut) [reply src/apps/monitoring/kadiska/mode/alerts.pm:96]: Yes it's done at the first set assignment

**Review by sdepassio — DISMISSED** [review]

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+        { label => 'rules-nodata-total-count', nlabel => 'rules.nodata.count', set => {
+                key_values => [ { name => 'total_nodata' }  ],
+                output_template => 'total no data: %s',
+                perfdatas => [ { template => '%d', min => 0 } ]
+            }
+        }
+    ];
+
+    $self->{maps_counters}->{rules} = [
+        { label => 'rule-ok-count',
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:77]: the nlabel is missing: ```rule.ok.count```

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+            set => {
+                key_values => [ { name => 'ok_count' } ],
+                output_template => 'ok count: %s',
+                display_ok => 0,
+                closure_custom_perfdata => sub { return 0; },
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:82]: you need to remove the closure_custom_perfdata

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+                perfdatas => [ {template => '%s', min => 0, label_extra_instance => 1, instance_use => 'name' } ]
+            }
+        },
+        { label => 'rule-critical-count',
+            set => {
+                key_values => [ { name => 'critical_count' } ],
+                output_template => 'critical count: %s',
+                display_ok => 0,
+                closure_custom_perfdata => sub { return 0; },
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:100]: the nlabel is missing: ```rule.criticals.count```

you need to remove the closure_custom_perfdata

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+                output_template => 'ok count: %s',
+                display_ok => 0,
+                closure_custom_perfdata => sub { return 0; },
+                perfdatas => [ {template => '%s', min => 0, label_extra_instance => 1, instance_use => 'name' } ]
+            }
+        },
+        { label => 'rule-warning-count',
+            set => {
+                key_values => [ { name => 'warning_count' } ],
+                output_template => 'warning count: %s',
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:89]: the nlabel is missing: ```rule.warnings.count```

you need to remove the closure_custom_perfdata

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+                display_ok => 0,
+                closure_custom_perfdata => sub { return 0; },
+                perfdatas => [ {template => '%s', min => 0, label_extra_instance => 1, instance_use => 'name' } ]
+            }
+        },
+        { label => 'rule-nodata-count',
+            set => {
+                key_values => [ { name => 'nodata_count' } ],
+                output_template => 'no data count: %s',
+                display_ok => 0,
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:108]: the nlabel is missing: ```rule.nodata.count```

you need to remove the closure_custom_perfdata

**Review by garnier-quentin — COMMENTED** [review]
same comments for the 4

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+            }
+        },
+        { label => 'rule-warning-count', nlabel => 'rule-warning-count', set => {
+                key_values => [ { name => 'warning_count' } ],
+                output_template => 'warning count: %s',
+                display_ok => 0,
+                perfdatas => [ {template => '%s', min => 0, label_extra_instance => 1, instance_use => 'name' } ]
+            }
+        },
+        { label => 'rule-critical-count', nlabel => 'rule-critical-count', set => {
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:91]: The syntax for perfdata is dot: ```rule.critical.count```

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+                key_values => [ { name => 'critical_count' } ],
+                output_template => 'critical count: %s',
+                display_ok => 0,
+                perfdatas => [ {template => '%s', min => 0, label_extra_instance => 1, instance_use => 'name' } ]
+            }
+        },
+        { label => 'rule-nodata-count', nlabel => 'rule-nodata-count', set => {
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:98]: The syntax for perfdata is dot: rule.nodata.count

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+        { label => 'rules-nodata-total-count', nlabel => 'rules.nodata.count', set => {
+                key_values => [ { name => 'total_nodata' }  ],
+                output_template => 'total no data: %s',
+                perfdatas => [ { template => '%d', min => 0 } ]
+            }
+        }
+    ];
+
+    $self->{maps_counters}->{rules} = [
+        { label => 'rule-ok-count', nlabel => 'rule-ok-count', set => {
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:77]: The syntax for perfdata is dot: rule.ok.count

File: src/apps/monitoring/kadiska/mode/alerts.pm [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+                key_values => [ { name => 'ok_count' } ],
+                output_template => 'ok count: %s',
+                display_ok => 0,
+                perfdatas => [ {template => '%s', min => 0, label_extra_instance => 1, instance_use => 'name' } ]
+            }
+        },
+        { label => 'rule-warning-count', nlabel => 'rule-warning-count', set => {
```
Comment (garnier-quentin) [inline src/apps/monitoring/kadiska/mode/alerts.pm:84]: The syntax for perfdata is dot: rule.warning.count

**Review by garnier-quentin — APPROVED** [review]

**Review by sdepassio — APPROVED** [review]

**Pull request merged**